bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/values/NumberValue.java

```diff
@@ -90,49 +92,54 @@ public String getMQTTpublishValue(@Nullable String pattern) {
 
     @Override
     public void update(Command command) throws IllegalArgumentException {
-        DecimalType oldvalue = (state == UnDefType.UNDEF) ? new DecimalType() : (DecimalType) state;
         BigDecimal newValue = null;
         if (command instanceof DecimalType) {
-            if (!checkConditions(((DecimalType) command).toBigDecimal(), oldvalue)) {
-                return;
-            }
-            state = (DecimalType) command;
+            newValue = ((DecimalType) command).toBigDecimal();
         } else if (command instanceof IncreaseDecreaseType || command instanceof UpDownType) {
+            BigDecimal oldValue = getOldValue();
             if (command == IncreaseDecreaseType.INCREASE || command == UpDownType.UP) {
-                newValue = oldvalue.toBigDecimal().add(step);
+                newValue = oldValue.add(step);
             } else {
-                newValue = oldvalue.toBigDecimal().subtract(step);
-            }
-            if (!checkConditions(newValue, oldvalue)) {
-                return;
+                newValue = oldValue.subtract(step);
             }
-            state = new DecimalType(newValue);
         } else if (command instanceof QuantityType<?>) {
-            QuantityType<?> qType = (QuantityType<?>) command;
-
-            if (qType.getUnit().isCompatible(Units.ONE)) {
-                newValue = qType.toBigDecimal();
-            } else {
-                qType = qType.toUnit(unit);
-                if (qType != null) {
-                    newValue = qType.toBigDecimal();
-                }
-            }
-            if (newValue != null) {
-                if (!checkConditions(newValue, oldvalue)) {
-                    return;
-                }
-                state = new DecimalType(newValue);
-            }
+            newValue = getQuantityTypeAsDecimal((QuantityType<?>) command);
         } else {
             newValue = new BigDecimal(command.toString());
-            if (!checkConditions(newValue, oldvalue)) {
-                return;
-            }
+        }
+        if (!checkConditions(newValue)) {
+            return;
+        }
+        // items with units specified in the label in the UI but no unit on mqtt are stored as
+        // DecimalType to avoid conversions (e.g. % expects 0-1 rather than 0-100)
+        if (!unit.equals(Units.ONE)) {
```

Comment:
Suggest to have constant on left side to eliminate any risk of NPE now or in the future:
```suggestion
        if (!Units.ONE.equals(unit)) {
```

Reply:
🤦 fixed